core/src/main/java/org/infinispan/commands/write/PutKeyValueCommand.java

```diff
@@ -96,13 +97,17 @@ public Object perform(InvocationContext ctx) throws Throwable {
       }
       MVCCEntry e = (MVCCEntry) ctx.lookupEntry(key);
 
-      //possible as in certain situations (e.g. when locking delegation is used) we don't wrap
-      if (e == null) return null;
+      if (e == null) {
+         throw new IllegalStateException("Not wrapped");
+      }
 
-      Object entryValue = e.getValue();
-      if (!valueMatcher.matches(e, null, value, valueEquivalence)) {
+      Object prevValue = e.getValue();
+      if (!valueMatcher.matches(prevValue, null, value, valueEquivalence)) {
          successful = false;
-         return entryValue;
+         return prevValue;
+      } else if (hasFlag(Flag.COMMAND_RETRY) && putIfAbsent) {
+         // TODO: hack branch!
+         e.setValue(null);
```

Comment:
All right, it's half-way fix, and it does not break the testsuite, so removing that branch...